lib/CL/devices/vulkan/pocl-vulkan.c

```diff
@@ -0,0 +1,3046 @@
+  d->cmd_buf_begin_info.pInheritanceInfo = NULL;
+
+  VkDescriptorPoolSize descriptor_pool_size[2]
+      = { { VK_DESCRIPTOR_TYPE_STORAGE_BUFFER, MAX_BUF_DESCRIPTORS },
+          { VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER, MAX_UBO_DESCRIPTORS } };
+
+  VkDescriptorPoolCreateInfo descriptor_pool_create_info
+      = { VK_STRUCTURE_TYPE_DESCRIPTOR_POOL_CREATE_INFO,
+          NULL,
+          VK_DESCRIPTOR_POOL_CREATE_FREE_DESCRIPTOR_SET_BIT,
+          MAX_DESC_SETS, /* maxSets */
+          2,             /* size of next array */
+          descriptor_pool_size };
+
+  VULKAN_CHECK (vkCreateDescriptorPool ( d->device,
+                                         &descriptor_pool_create_info,
+                                         NULL, &d->buf_descriptor_pool));
+
+  POCL_INIT_COND (d->wakeup_cond);
+
+  POCL_FAST_INIT (d->wq_lock_fast);
+
+  d->work_queue = NULL;
+
+  pthread_create (&d->driver_pthread_id, NULL, pocl_vulkan_driver_pthread,
+                  dev);
+
+  return CL_SUCCESS;
+}
+
+#if 0
+/* TODO finish implementation */
+cl_int
+pocl_vulkan_uninit (unsigned j, cl_device_id device)
+{
+  pocl_vulkan_device_data_t *d = (pocl_vulkan_device_data_t*)device->data;
+
+  vkDestroyCommandPool (d->device, d->command_pool, NULL);
+
+  vkDestroyDescriptorPool (d->device, d->buf_descriptor_pool, NULL);
+
+  /* destroy logical device */
+  vkDestroyDevice (d->device, NULL);
+
+  return CL_SUCCESS;
+
+  /* TODO this must be called after all devices !! */
+  vkDestroyInstance(pocl_vulkan_instance, NULL);
+}
+
+cl_ulong
+pocl_vulkan_get_timer_value(void *data)
+{
+  return pocl_gettimemono_ns();
+}
+#endif
+
+int
+run_and_append_output_to_build_log (cl_program program, char *const *args)
+{
+  int errcode = CL_SUCCESS;
+
+  char *capture_string = NULL;
+
+  capture_string = malloc (MAIN_PROGRAM_LOG_SIZE);
+  POCL_GOTO_ERROR_ON ((capture_string == NULL), CL_OUT_OF_HOST_MEMORY,
+                      "Error while allocating temporary memory\n");
+
+  capture_string[0] = 0;
+  size_t log_len = strlen (program->main_build_log);
+  /* if the log is full, run without capture */
+  if (log_len + 1 >= MAIN_PROGRAM_LOG_SIZE)
+    return pocl_run_command (args);
+
+  size_t capture_len = MAIN_PROGRAM_LOG_SIZE - log_len - 1;
+  errcode
+      = pocl_run_command_capture_output (capture_string, &capture_len, args);
+  if (capture_len > 0)
+    {
+      assert (log_len + capture_len < MAIN_PROGRAM_LOG_SIZE - 1);
+      memcpy (program->main_build_log + log_len, capture_string, capture_len);
+      program->main_build_log[log_len + capture_len] = 0;
+    }
+
+ERROR:
+  free (capture_string);
+  return errcode;
+}
+
+int
+pocl_vulkan_build_source (cl_program program, cl_uint device_i,
+                          cl_uint num_input_headers,
+                          const cl_program *input_headers,
+                          const char **header_include_names, int link_program)
+{
+#ifdef HAVE_CLSPV
```

Comment:
Why #ifdef here? What if we don't have CLSPV, is the Vulkan driver useful for something then?

Reply:
yes, you can load and run (using `clCreateProgramWithBinary()` ) programs built on another machine with pocl-vulkan, similarly to CPU drivers.